tracing: add support for OpenCensus

brianwolfe opened the pull request:
Add support for OpenCensus tracing. This is similar to the existing
Stackdriver support. This is tested by adding a new tracing bootstrap
golden test comparison and by individually testing the configuration
options, similar to the existing tracers.

This implements the API from this PR:
https://github.com/istio/api/pull/1563

After this lands, I will update the documentation for tracing for OpenCensusAgent and in general for context propagation, since the existing documentation is a bit confusing about which headers need to be propagated in which cases.

And to help us figure out who should review this PR, please 
put an X in all the areas that this PR affects.

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[ ] Networking
[ ] Performance and Scalability
[X] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure


Pull Request Attributes

Please check any characteristics that apply to this pull request. 

[ ] Does not have any changes that may affect Istio users.

Commit:
tracing: add support for OpenCensus

Add support for OpenCensus tracing. This is similar to the existing
Stackdriver support. This is tested by adding a new tracing bootstrap
golden test comparison and by individually testing the configuration
options, similar to the existing tracers.

This implements the API from this PR:
https://github.com/istio/api/pull/1563

Change-Id: I65d87c6c5e3e52804b06f6b3d0a35f9515e43c7e
Reviewed-on: https://gerrit.musta.ch/c/public/istio/+/559
Reviewed-by: Jungho Ahn <jungho.ahn@airbnb.com>

istio-testing commented:
@brianwolfe: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
release-notes_istio | 4a5b73f3ce19f37ef2941feff589e9549789753c | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/27019/release-notes_istio/2855) | `/test release-notes_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

Commit:
release-notes: add note for opencensusagent

brianwolfe commented:
@nrjpoddar, @douglas-reid PTAL.

douglas-reid commented on pkg/bootstrap/option/convert.go:
```diff
@@ -180,6 +181,32 @@ func durationConverter(value *types.Duration) convertFunc {
+				envoyContexts = append(envoyContexts, "CLOUD_TRACE_CONTEXT")
+			case meshAPI.Tracing_OpenCensusAgent_B3:
+				envoyContexts = append(envoyContexts, "B3")
+				// Ignore UNSPECIFIED
```
super minor nit: i might move this comment up to the top above W3_TRACE_CONTEXT, as the indent makes it a bit funky at the bottom.

brianwolfe replied:
thanks, that looks better now.

douglas-reid commented on pkg/bootstrap/option/convert.go:
```diff
@@ -180,6 +181,32 @@ func durationConverter(value *types.Duration) convertFunc {
+func openCensusAgentContextConverter(contexts []meshAPI.Tracing_OpenCensusAgent_TraceContext) convertFunc {
+	return func(*instance) (interface{}, error) {
+		if len(contexts) == 0 {
+			return convertToJSON([]string{"TRACE_CONTEXT", "GRPC_BIN", "CLOUD_TRACE_CONTEXT", "B3"}), nil
```
probably doesn't make a difference, but should we just have this be a preconverted const or var (something like AllContexts) ?

brianwolfe replied:
SG. I wanted to keep the scope of envoy trace contexts to just this function so it's easy to extract out/modify when the new telemetry api starts to land. Made it into a pre-converted var in the function.

Commit:
tests/integration: add opencensusagent trace test

This test adds integration support for opencensusagent, sharing most of
the work from the existing zipkin integration tests, this exports traces
to an opentelemetry-collector instance, which forwards the traces to
zipkin. This same setup should be usable when envoy supports
opentelemetry tracing natively.

brianwolfe commented:
@howardjohn PTAL

mandarjog commented on pkg/bootstrap/option/convert.go:
```diff
@@ -180,6 +181,33 @@ func durationConverter(value *types.Duration) convertFunc {
+// openCensusAgentContextConverter returns a converter that returns the list of
+// distributed trace contexts to propagate with envoy.
+func openCensusAgentContextConverter(contexts []meshAPI.Tracing_OpenCensusAgent_TraceContext) convertFunc {
+	allContexts := "[\"TRACE_CONTEXT\",\"GRPC_BIN\",\"CLOUD_TRACE_CONTEXT\",\"B3\"]"
```
Nit: use back quotes

brianwolfe replied:
Done

mandarjog commented on pkg/bootstrap/option/instances_test.go:
```diff
@@ -204,6 +204,37 @@ func TestOptions(t *testing.T) {
+				meshAPI.Tracing_OpenCensusAgent_GRPC_BIN,
+				meshAPI.Tracing_OpenCensusAgent_W3C_TRACE_CONTEXT,
+			}),
+			expected: "[\"CLOUD_TRACE_CONTEXT\",\"B3\",\"GRPC_BIN\",\"TRACE_CONTEXT\"]",
```
Sorry missed this earlier, but you could use back quotes everywhere.

brianwolfe replied:
Done. You are right, I should have looked for all usage when you made that comment. Thanks!

mandarjog commented on pkg/bootstrap/testdata/xdsproxy_golden.json:
```diff
@@ -249,9 +249,11 @@
           "envoy_grpc": {
             "cluster_name": "xds-grpc"
           }
+        
```
Why is this change here?

brianwolfe replied:
This is unrelated. It looks like a previous change did not actually update using the standard mechanism of:
`REFRESH_GOLDEN=true go test ./pkg/bootstrap/...`
I can remove it, but it will need to be updated eventually anyway to make it easier to actually make changes to the template.

brianwolfe replied:
I backed this change out, I'll submit seperately.

mandarjog reviewed (commented):
Can we add an e2e test here ?
Take a look At the stackdriver test that uses fake endpoint.
@gargnupur can you point him to it?

nrjpoddar commented on pkg/bootstrap/option/convert.go:
```diff
@@ -180,6 +181,33 @@ func durationConverter(value *types.Duration) convertFunc {
+			case meshAPI.Tracing_OpenCensusAgent_CLOUD_TRACE_CONTEXT:
+				envoyContexts = append(envoyContexts, "CLOUD_TRACE_CONTEXT")
+			case meshAPI.Tracing_OpenCensusAgent_B3:
+				envoyContexts = append(envoyContexts, "B3")
```
Should there be a default error case in switch or by construction there cannot be values other than the ones you’re switching on?

brianwolfe replied:
By construction there cannot be more, aside from Tracing_OpenCensusAgent_UNSPECIFIED, which is intentionally skipped.

nrjpoddar commented on releasenotes/notes/opencensusagent.yaml:
```diff
@@ -0,0 +1,7 @@
+
+releaseNotes:
+  - |
+    **Updated** added support for OpenCensusAgent formatted trace export with configurable trace context headers.
```
Should we add the new tracers that are now supported explicitly?

brianwolfe replied:
Do you mean the contexts that are supported? This only adds a single tracer w/ multiple contexts propagation header options.

nrjpoddar replied:
I see. I think mentioning OpenCensus here and a doc for contexts make sense.

gargnupur commented:
> Can we add an e2e test here ?
> Take a look At the stackdriver test that uses fake endpoint.
> @gargnupur can you point him to it?

Stackdriver tracing test: https://github.com/istio/istio/blob/master/tests/integration/telemetry/stackdriver/stackdriver_filter_test.go#L330:1
Zipkin tracing test:
https://github.com/istio/istio/blob/master/tests/integration/telemetry/tracing/servertracing/tracing_test.go

gargnupur commented on tools/packaging/common/envoy_bootstrap.json:
```diff
@@ -627,6 +627,29 @@
+          "constant_sampler": {
+            "decision": "ALWAYS_PARENT"
+          },
+          "max_number_of_annotations": 200,
```
should this be configurable?

brianwolfe replied:
@nrjpoddar and I discussed in the API PR and I ended up removing the tuning vars from the API. If you think it is worthwhile, I can add env variable tuning of these, but it doesn't appear to have a real use-case by users.

https://github.com/istio/api/pull/1591

istio-testing commented:
@brianwolfe: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integ-multicluster-k8s-tests_istio | 3d897f6edcbb7094e7766ba1ff99defdc32e10ad | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/27019/integ-multicluster-k8s-tests_istio/10079) | `/test integ-multicluster-k8s-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

brianwolfe commented:
I'll work on adding the e2e test.

brianwolfe commented:
FYI, first integration test attempt, so I ran into some issues. Got it going now though: https://github.com/istio/istio/issues/27061

istio-testing commented:
@brianwolfe: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integ-multicluster-k8s-tests_istio | 3d897f6edcbb7094e7766ba1ff99defdc32e10ad | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/27019/integ-multicluster-k8s-tests_istio/10079) | `/test integ-multicluster-k8s-tests_istio`
lint_istio | 4ef0af5ac9e028a283ce4eec4cb7895dbf89332e | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/27019/lint_istio/19796) | `/test lint_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@brianwolfe: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integ-multicluster-k8s-tests_istio | 3d897f6edcbb7094e7766ba1ff99defdc32e10ad | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/27019/integ-multicluster-k8s-tests_istio/10079) | `/test integ-multicluster-k8s-tests_istio`
lint_istio | 61905b9f81cf32ff41ae5128a6f6c5fd683a2ad6 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/27019/lint_istio/19798) | `/test lint_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@brianwolfe: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integ-multicluster-k8s-tests_istio | 3d897f6edcbb7094e7766ba1ff99defdc32e10ad | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/27019/integ-multicluster-k8s-tests_istio/10079) | `/test integ-multicluster-k8s-tests_istio`
lint_istio | a53d9ca647632fcf169f81f32678d5bf2a3db147 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/27019/lint_istio/19799) | `/test lint_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

brianwolfe commented:
/test lint_istio

istio-testing commented:
@brianwolfe: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integ-multicluster-k8s-tests_istio | 3d897f6edcbb7094e7766ba1ff99defdc32e10ad | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/27019/integ-multicluster-k8s-tests_istio/10079) | `/test integ-multicluster-k8s-tests_istio`
lint_istio | aa9760880ef4a06695014da6408e858bac90e062 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/27019/lint_istio/19800) | `/test lint_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

brianwolfe commented:
@gargnupur Could you take a look at what I did here to try and test the OpenCensusAgent trace exporter? It's my first try at doing an e2e test in this framework. I think it makes sense, but I'm not positive.

I configured it to do: proxy -> opentelemetry-collector -> zipkin so we can reuse the zipkin querying for opencensusagent traces.

brianwolfe commented:
@mandarjog is this similar to what you had in mind for an e2e test?

bianpengyuan commented on pkg/test/framework/components/opentelemetry/opentelemetry-collector.yaml:
```diff
@@ -0,0 +1,122 @@
+kind: ConfigMap
+metadata:
+  name: opentelemetry-collector
+  namespace: istio-system
```
Should remove istio-system namespace from here? The component logic decides which telemetry namespace this file is applied to.

brianwolfe replied:
Done

bianpengyuan reviewed (commented):
The new tracing test lgtm

brianwolfe commented on releasenotes/notes/opencensusagent.yaml:
```diff
@@ -0,0 +1,7 @@
+
+releaseNotes:
+  - |
+    **Updated** added support for OpenCensusAgent formatted trace export with configurable trace context headers.
```
Done.

brianwolfe commented:
@howardjohn can you take a look for environments or let me know someone better to review?
